src/main/java/net/minestom/server/utils/block/BlockBreakCalculation.java

```diff
@@ -0,0 +1,192 @@
+import net.minestom.server.entity.GameMode;
+import net.minestom.server.entity.Player;
+import net.minestom.server.entity.attribute.Attribute;
+import net.minestom.server.gamedata.tags.Tag;
+import net.minestom.server.gamedata.tags.Tag.BasicType;
+import net.minestom.server.instance.Instance;
+import net.minestom.server.instance.block.Block;
+import net.minestom.server.item.ItemComponent;
+import net.minestom.server.item.ItemStack;
+import net.minestom.server.item.component.EnchantmentList;
+import net.minestom.server.item.component.Tool;
+import net.minestom.server.item.enchant.Enchantment;
+import net.minestom.server.potion.PotionEffect;
+import net.minestom.server.registry.Registry;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Objects;
+
+public class BlockBreakCalculation {
+
+    public static final int UNBREAKABLE = -1;
+    private static final Tag WATER_TAG = Objects.requireNonNull(MinecraftServer.getTagManager().getTag(BasicType.FLUIDS, "minecraft:water"));
+    // The vanilla client checks for bamboo breaking speed with item instanceof SwordItem.
+    // We could either check all sword ID's, or the sword tag.
+    // Since tags are immutable, checking the tag seems easier to understand
+    private static final Tag SWORD_TAG = Objects.requireNonNull(MinecraftServer.getTagManager().getTag(BasicType.ITEMS, "minecraft:swords"));
+
+    /**
+     * Calculates the block break time in ticks
+     *
+     * @return the block break time in ticks, -1 if the block is unbreakable
+     */
+    public static int breakTicks(@NotNull Block block, @NotNull Player player) {
+        if (player.getGameMode() == GameMode.CREATIVE) {
+            // Creative can always break blocks instantly
+            return 0;
+        }
+        // Taken from minecraft wiki Breaking#Calculation
+        // More information to mimic calculations taken from minecraft's source
+        Registry.BlockEntry registry = block.registry();
+        double blockHardness = registry.hardness();
+        if (blockHardness == -1) {
+            // Bedrock, barrier, and unbreakable blocks
+            return UNBREAKABLE;
+        }
+        ItemStack item = player.getItemInMainHand();
+        // Bamboo is hard-coded in client
+        if (block.id() == Block.BAMBOO.id() || block.id() == Block.BAMBOO_SAPLING.id()) {
+            if (SWORD_TAG.contains(item.material().namespace())) {
+                return 0;
+            }
+        }
+        Tool tool = item.get(ItemComponent.TOOL);
+        boolean isBestTool = canBreakBlock(tool, block);
+        float speedMultiplier;
+
+        if (isBestTool) {
+            speedMultiplier = getMiningSpeed(tool, block);
+
+            // Also, the wiki's canBreak seems to be this >1
+            if (speedMultiplier > 1F) {
+                // Wiki tells us about calculations for efficiency, but enchantments don't apply these calculations in minestom.
```

Comment:
Since data driven enchantments, efficiency uses the mining efficiency attribute anyway so this is correct (though minestom doesn't apply currently)

Reply:
Should I change comment? Tried to explain deviation from wiki, in case someone looks at it in the future